### DISCUSSION TITLE
Renaming of Node methods causes problems in transpilation

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#83] [Ideas]
The Java parsers generated by CongoCC have full type information because of how CongoCC is integrated with Java, but the polyglot versions don't. So with `Node` methods renamed from e.g. `getChild` -> `get` and `setChild` -> `set`, this causes problems because those method names are also used by various collections in the Java runtime. Transpiling `get` for collections, for example, in both Python and C#, involves replacing `get(X)` with `[X]` - however, this only works for collections! Now if we had full type information at the transpilation stage, we might figure this out, but it's potentially a lot of work to do  just to deal with ambiguities caused by lack of type information. Could we go back to `getChild()`, `setChild()` etc. and do away with the newer methods? In general if we are going to make polyglot support easy, any and all method renaming and class hierarchy/interface implementation changes should be reviewed for impact before just making the changes. I attempted to switch over to `get()`/`set()` etc. when doing the lexer refactoring, and have run into this with some of the C# injections.

**Comment by revusky** [comment]
>this causes problems because those method names are also used by various collections in the Java runtime.

Well, yes, that's true, but I had been thinking that this was something positive! I mean, the fact that any `Node`  now also implements List<Node> means that certain things can be expressed much more elegantly than before. At least on the Java side. But I tended to think (though probably vaguely) that this advantage would carry over to other languages, that we could just manipulate/access Node instances in a more idiomatic way typically. 

But I do have to admit that I am not sure what the problem really is with the method being called get() as opposed to getChild(). My tendency would be to think that if we can write node[i] in CSharp, say, instead of node.get(i) in Java, say, that is all the better, no?

One perhaps wooly-minded thought I was having today was I was thinking that if there is insufficient information, maybe we could have some optional annotations that could resolve whatever ambiguities that exist wrt transpilation. I mean, the Java code itself could carry hints, via some annotations, regarding how a certain construct shoud be transpiled into Python or CSharp. I suppose you know that currently, within a Java code injection, you can add @Property annotation to a field declaration and it generates getter/setters for that field. And it's really a very trivial thing. Just look at: https://github.com/congo-cc/congo-parser-generator/blob/main/src/java/org/congocc/codegen/java/JavaCodeUtils.java#L22 

It really took very little code to implement that. 

Do you think that is a crazy idea? I mean to leverage annotations in a broadly analogous way to add information for transpilation...

Meanwhile, the machinery does exist to have the ability to directly inject code in Python or C#, since we have the ability to parse python and Csharp directly and for the code to live in the same AST. But somebody (I guess me, but it would be great if it wasn't me!) has to buckle down and get that working.

By the way, did you see the newer terse syntax in FreeMarker? https://parsers.org/announcements/not-your-fathers-freemarker/

**Comment by vsajip** [comment]
>  I am not sure what the problem really is with the method being called get()

With the limited type information available to the transpiler implementations, they can't always be sure when `get()` usage is for a collection method and when it is for something else like a `Node`. Collection methods have different notations in e.g. C# where you use `X[Y]` but a call to an arbitrary `get()` method on a non-collection instance would need to be transpiled as `X.Get(Y)`. If you see a `x.get(y)` in an injection, it's hard to know what exactly that is and how to transpile it.

> we could have some optional annotations that could resolve whatever ambiguities that exist wrt transpilation ... Do you think that is a crazy idea?

We probably need to decide between that approach and another we discussed on Slack, such as having semantic actions in a DSL that then was uniformly transpilable to the target language.

> did you see the newer terse syntax in FreeMarker

Yes, I did. I had to merge your changes to `lexer.py.ftl`, which I was working on.

**Reply by revusky** [reply]
>> did you see the newer terse syntax in FreeMarker

>Yes, I did. I had to merge your changes to lexer.py.ftl, which I was working on.

One thing that occurred to me shortly after implementing this is the possibility of enforcing (maybe optionally) correct indentation *à la Python*. Since the terse directives all start on their own line, we can easily check for correct indentation. So we either just warn on something like:

          #if foo
                #if bar
          /#if <-- wrong dedent!
        /#if

 which is not indented properly, we either (a) complain or (b) create an outright impedance, like with Python. I am tending towards experimenting with (b). That could be a bit annoying maybe, but it also would result in templates being more readable typically than without it.

What say you?

**Reply by vsajip** [reply]
No objections from me, though definitely it should have a way of turning the checks off. I personally prefer to indent with two spaces, though only within the body of a macro or template - you'll see what I mean if you look at the Python/C# templates I've written. Maybe there should be a "linting" option you can run standalone which just prints out any "offenses".

**Reply by adMartem** [reply]
Wouldn't it be insensitive to the actual indentation amount?  In other words, the terminator would have to match the initiator indentation, not any particular number of spaces, right?  If so, I think enforcing it would always be fine (unless I'm missing something).

**Comment by revusky** [comment]
> > I am not sure what the problem really is with the method being called get()
> 
> With the limited type information available to the transpiler implementations, they can't always be sure when `get()` usage is for a collection method and when it is for something else like a `Node`.

Well, in Java, as long as your class implements `java.util.List` or `java.util.Map`, I guess it's effectively a collection. I mean, in principle (at least AFAICS) there is nothing special about `java.util.ArrayList` or `java.util.HashMap`. In principle, anyway... OTOH, `java.lang.String` lives in `java.lang` and is arguably more of a core type. Well, it doesn't need to be explicitly imported, for one thing. It also has a bit of syntactic sugar, that you can use the `+` operator and so on. As well as syntax for a string literal. I also have little doubt that there are quite a few special hacks inside the JVM to deal with strings specially here and there. Also the same applies to primitive wrappers, like `java.lang.Integer` and so on. 

That said, I have sometimes wondered whether things like `ArrayList` and `HashMap` aren't so similarly basic that they should also be auto-imported and have a bit of extra syntactic sugar around them, like being able to write list[i] instead of list.get(i) or have a terse syntax for list and hash literals...(AFAICS, the Java language designers have been surprisingly stodgy in this regard. Even regarding strings, you still have to write `string.charAt(i)` instead of the more natural `string[i]`). But obviously, other languages have been more aggressive about dealing with these things in a more natural syntax...

But anyway, my current thinking is that even if Java doesn't provide this, and you still have to write `node.get(i)` when it would be natural/aesthetic to be able to write `node[i]`, when we're generating code in a language that has this, we should preferably have this syntactic sugar working there!

Now, as for ambiguities that arise, having slept on this, I tend to think that my idea of using annotations in whatever cases it is necessary to disambiguate looks feasible. Roughly, I'm thinking about stuff like this:

          Annotation ann = methodDeclaration.firstChildOfType(Annotation.class, ann->ann.getName().equals("TranspileAsProperty");
          if (ann != null) {
              do something a bit different here maybe...
          } else {
              do whatever the default is without the hint
          }

And then you could have in your code injection something like:

        @TranspileAsProperty("python", "csharp")
        public Something getSomething() {...}

But the annotation is just an optional hint that you can place to help the transpilation machinery transpile some element, no?

But really, I mean to say, we have all the machinery in place to handle this sort of thing pretty easily, we have a full AST and we can easily fish out whatever annotations are there. So we might as well take advantage of that fact, no?

I mean, given the fact that Java allows you to put annotations practically anywhere in your code and we have all the machinery to parse this and access it from our client code...









> Collection methods have different notations in e.g. C# where you use `X[Y]` but a call to an arbitrary `get()` method on a non-collection instance would need to be transpiled as `X.Get(Y)`. If you see a `x.get(y)` in an injection, it's hard to know what exactly that is and how to transpile it.
> 
> > we could have some optional annotations that could resolve whatever ambiguities that exist wrt transpilation ... Do you think that is a crazy idea?
> 
> We probably need to decide between that approach and another we discussed on Slack, such as having semantic actions in a DSL that then was uniformly transpilable to the target language.
> 
> > did you see the newer terse syntax in FreeMarker
> 
> Yes, I did. I had to merge your changes to `lexer.py.ftl`, which I was working on.
